pmlc/target/x86/xsmm_lowering.cc

```diff
@@ -0,0 +1,385 @@
+
+  PatternMatchResult matchAndRewrite(xsmm::GemmOp op,
+                                     PatternRewriter &rewriter) const override {
+    Impl impl(op, rewriter);
+    auto symbol = impl.getOrInsertFunc();
+    auto a = impl.prepareOperand(op.a(), op.aAccessMap(), op.getOperandsForA(),
+                                 op.aTileMap(), kDimensionM);
+    auto b = impl.prepareOperand(op.b(), op.bAccessMap(), op.getOperandsForB(),
+                                 op.bTileMap(), kDimensionN);
+    auto c = impl.prepareOperand(op.c(), op.cAccessMap(), op.getOperandsForC(),
+                                 op.cTileMap(), kDimensionN);
+    SmallVector<Value, 9> args{a.memref,           b.memref,
+                               c.memref,           a.leadingDimStride,
+                               b.leadingDimStride, c.leadingDimStride};
+    for (auto i : impl.tile) {
+      args.push_back(impl.createConstantIntOp(i));
+    }
+    rewriter.replaceOpWithNewOp<CallOp>(op, symbol, ArrayRef<Type>{}, args);
+    return matchSuccess();
+  }
+
+  struct Impl {
+    xsmm::GemmOp op;
+    PatternRewriter &rewriter;
+    Location loc;
+    ModuleOp module;
+    Type i32Type;
+    Type elementType;
+    UnrankedMemRefType unrankedType;
+    SmallVector<unsigned, 3> tile;
+
+    struct PreparedOperand {
+      Value memref;
+      Value leadingDimStride;
+    };
+
+    Impl(xsmm::GemmOp op, PatternRewriter &rewriter)
+        : op(op), rewriter(rewriter), loc(op.getLoc()),
+          module(op.getParentOfType<ModuleOp>()),
+          i32Type(rewriter.getIntegerType(32)),
+          elementType(rewriter.getF32Type()),
+          unrankedType(
+              UnrankedMemRefType::get(elementType, /*memorySpace=*/0)) {
+      for (auto attr : op.tile().getValue()) {
+        tile.push_back(attr.cast<IntegerAttr>().getInt());
+      }
+    }
+
+    FlatSymbolRefAttr getOrInsertFunc() {
+      const char *symbol = "plaidml_rt_xsmm_gemm_f32";
+      auto context = module.getContext();
+      if (module.lookupSymbol(symbol)) {
+        return SymbolRefAttr::get(symbol, context);
+      }
+      OpBuilder builder(module.getBodyRegion());
+      std::array<Type, 9> inputs{unrankedType, unrankedType, unrankedType,
+                                 i32Type,      i32Type,      i32Type,
+                                 i32Type,      i32Type,      i32Type};
+      ArrayRef<Type> results{};
+      auto funcType = builder.getFunctionType(inputs, results);
+      ArrayRef<NamedAttribute> attrs{};
+      builder.create<FuncOp>(loc, symbol, funcType, attrs);
+      return SymbolRefAttr::get(symbol, context);
+    }
+
+    PreparedOperand prepareOperand(Value operand, AffineMap accessMap,
+                                   ValueRange mapOperands, AffineMap tileMap,
+                                   unsigned leadingDimPos) {
+      ArrayRef<Value> empty{};
+      auto offsets = expandAffineMap(rewriter, loc, accessMap, mapOperands);
```

Comment:
I'm think there might be an easier way to do this, not a blocker for now, although it brings in a lot of copied code.

Reply:
My plan is to make this code accessible upstream so that we don't need to duplicate it. But this requires a patch upstream.